src/PostgREST/Metrics.hs

```diff
@@ -0,0 +1,35 @@
+{-# LANGUAGE NamedFieldPuns #-}
+module PostgREST.Metrics
+  ( init
+  , MetricsState
+  , observationMetrics
+  , metricsToText
+  ) where
+
+import qualified Data.ByteString.Lazy as LBS
+
+import qualified Prometheus as Prom
+
+import PostgREST.Observation
+
+import Protolude
+
+newtype MetricsState = MetricsState
+  { counterPoolTimeout          :: Prom.Counter
+  }
+
+init :: IO MetricsState
+init = do
+  counter <- Prom.register $ Prom.counter (Prom.Info "pgrst_pool_timeouts" "The number of connection timeouts in the pool")
+  pure $ MetricsState counter
+
+observationMetrics :: MetricsState -> ObservationHandler
+observationMetrics MetricsState{counterPoolTimeout} obs = case obs of
+  (PoolAcqTimeoutObs _) -> do
+    Prom.incCounter counterPoolTimeout
+  _ ->
+    pure ()
```

Comment:
Implementation came out quite nice, just based on observations. This way the metric collection don't need to be scattered across the codebase.